core-rust/state-manager/src/state_manager.rs

```diff
@@ -178,6 +174,74 @@ impl<S> StateManager<S> {
             prometheus_registry,
         }
     }
+
+    fn execute_and_commit_validator_transaction<T: SubstateStore>(
+        scrypto_interpreter: &ScryptoInterpreter<DefaultWasmEngine>,
+        fee_reserve_config: &FeeReserveConfig,
+        execution_config: &ExecutionConfig,
+        validator_transaction: ValidatorTransaction,
+        substate_store: &mut T,
+        hash_tree: &mut HashTree
+    ) -> Result<LedgerTransaction, String> {
+        let prepared_txn = validator_transaction.prepare();
+        let executable = &prepared_txn.get_executable();
+
+        let result = Self::execute_and_commit_transaction(
+            scrypto_interpreter,
+            fee_reserve_config,
+            execution_config,
+            executable,
+            substate_store,
+            hash_tree
+        );
+
+        match result {
+            TransactionResult::Commit(commit_result) => match commit_result.outcome {
+                TransactionOutcome::Failure(error) => {
+                    Err(format!("Validator txn failed: {:?}", error))
+                }
+                TransactionOutcome::Success(..) => {
+                    Ok(LedgerTransaction::Validator(validator_transaction))
+                }
+            },
+            TransactionResult::Reject(reject_result) => {
+                Err(format!("Validator txn rejected: {:?}", reject_result))
+            }
+        }
+    }
+
+    // DEV: question - should this, instead, become a new behavior of
+    // radix_engine::transaction::execute_and_commit_transaction() ?
```

Comment:
From the node perspective: `execute_and_commit_transaction` (I'd argue all `execute_*`) should not be used directly anyway, with some exceptions (i.e. `preview_call`) since you want to hit the execution cache first.
From the perspective of engine/scrypto: unless the hash root is actually used inside of it I don't see why it needs to know about it.

Reply:
I had an impression that `execute_and_commit_transaction()` is only a convenience method, while the "raw business" is the `execute_transaction()` itself.

For (unrelated) reasons, we want the JMT impl to live under the Scrypto repo (even though, technically, it could be fully contained within Node). And in this context, I assumed that the convenience `execute_and_commit_transaction()` could take the tree update responsibility into account as well - but that's just a thought. I'll probably only do it if it simplifies some code (of Node or of Scrypto or even of tests), I guess that's what "convenience method" means :)